engines/wintermute/base/gfx/opengl/base_render_opengl3d_shader.cpp

```diff
@@ -0,0 +1,617 @@
+
+	case Graphics::BLEND_ADDITIVE:
+		glBlendFunc(GL_SRC_ALPHA, GL_ONE);
+		break;
+
+	case Graphics::BLEND_SUBTRACTIVE:
+		// wme3d takes the color value here
+		glBlendFunc(GL_ZERO, GL_ONE_MINUS_SRC_COLOR);
+		break;
+
+	default:
+		error("BaseRenderOpenGL3DShader::setSpriteBlendMode unsupported blend mode %i", blendMode);
+	}
+}
+
+void BaseRenderOpenGL3DShader::setAmbientLight() {
+	byte a = RGBCOLGetA(_ambientLightColor);
+	byte r = RGBCOLGetR(_ambientLightColor);
+	byte g = RGBCOLGetG(_ambientLightColor);
+	byte b = RGBCOLGetB(_ambientLightColor);
+
+	if (!_overrideAmbientLightColor) {
+		uint32 color = _gameRef->getAmbientLightColor();
+
+		a = RGBCOLGetA(color);
+		r = RGBCOLGetR(color);
+		g = RGBCOLGetG(color);
+		b = RGBCOLGetB(color);
+	}
+
+	Math::Vector4d value;
+	value.x() = r / 255.0f;
+	value.y() = g / 255.0f;
+	value.z() = b / 255.0f;
+	value.w() = a / 255.0f;
+
+	_modelXShader->use();
+	_modelXShader->setUniform("ambientLight", value);
+}
+
+int BaseRenderOpenGL3DShader::maximumLightsCount() {
+	return 8;
+}
+
+void BaseRenderOpenGL3DShader::enableLight(int index) {
+	_modelXShader->use();
+	Common::String uniform = Common::String::format("lights[%i].enabled", index);
+	_modelXShader->setUniform1f(uniform.c_str(), 1.0f);
+}
+
+void BaseRenderOpenGL3DShader::disableLight(int index) {
+	_modelXShader->use();
+	Common::String uniform = Common::String::format("lights[%i].enabled", index);
+	_modelXShader->setUniform1f(uniform.c_str(), -1.0f);
+}
+
+void BaseRenderOpenGL3DShader::setLightParameters(int index, const Math::Vector3d &position, const Math::Vector3d &direction, const Math::Vector4d &diffuse, bool spotlight) {
+	Math::Vector4d position4d;
+	position4d.x() = position.x();
+	position4d.y() = position.y();
+	position4d.z() = position.z();
+	position4d.w() = 1.0f;
+
+	Math::Vector4d direction4d;
+	direction4d.x() = direction.x();
+	direction4d.y() = direction.y();
+	direction4d.z() = direction.z();
+	direction4d.w() = 0.0f;
+
+	if (spotlight) {
+		direction4d.w() = -1.0f;
+	}
+
+	_modelXShader->use();
+
+	Common::String uniform = Common::String::format("lights[%i]._position", index);
+	_modelXShader->setUniform(uniform.c_str(), position4d);
+
+	uniform = Common::String::format("lights[%i]._direction", index);
+	_modelXShader->setUniform(uniform.c_str(), direction4d);
+
+	uniform = Common::String::format("lights[%i]._color", index);
+	_modelXShader->setUniform(uniform.c_str(), diffuse);
+}
+
+bool BaseRenderOpenGL3DShader::enableShadows() {
+	warning("BaseRenderOpenGL3DShader::enableShadows not implemented yet");
+	return true;
+}
+
+bool BaseRenderOpenGL3DShader::disableShadows() {
+	warning("BaseRenderOpenGL3DShader::disableDhadows not implemented yet");
+	return true;
+}
+
+bool BaseRenderOpenGL3DShader::stencilSupported() {
+	// assume that we have a stencil buffer
+	return true;
+}
+
+BaseImage *BaseRenderOpenGL3DShader::takeScreenshot() {
+	warning("BaseRenderOpenGL3DShader::takeScreenshot not yet implemented");
```

Comment:
The screenshot functionality is used for savegame purposes, and thus not entirely unimportant.